plugins/filter_aws/aws.c

```diff
@@ -370,6 +400,208 @@ static int get_vpc_metadata(struct flb_filter_aws *ctx)
+    /* \n is separator, therefore number of items = number of \n + 1 */
+    ctx->tags_count = 1;
+    for (i = 0; i < len; i++) {
+        if (tags_list[i] == '\n') {
+            ctx->tags_count++;
+        }
+    }
+    ctx->tag_keys = flb_calloc(ctx->tags_count, sizeof(flb_sds_t*));
+    if (!ctx->tag_keys) {
+        flb_errno();
+        flb_sds_destroy(tags_list);
+        return -1;
+    }
+    ctx->tag_keys_len = flb_calloc(ctx->tags_count, sizeof(size_t*));
+    if (!ctx->tag_keys_len) {
+        flb_errno();
+        flb_sds_destroy(tags_list);
+        return -1;
+    }
+
+    /* go over the response and initialize tag_keys values */
+    /* code below finds two indices which define tag key and copies them to ctx */
+    while (tag_end <= len) {
+        /* replace \n with \0 to 'clearly' separate tag key strings */
+        if (tags_list[tag_end] == '\n') {
+            tags_list[tag_end] = '\0';
+        }
+        if ((tags_list[tag_end] == '\0' || tag_end == len) && (tag_start < tag_end)) {
+            /* length of tag key characters is the difference between start and end */
+            /* for instance, if tag name is 'Name\0...', the corresponding values are */
+            /*   tag_start = 0, points to 'N' */
+            /*   tag_end = 4, points to '\0' just after 'e' */
+            /*   f.e.: 4 - 0 = 4, which is equal to len("Name") */
+            tag_key_len = tag_end - tag_start;
+            ctx->tag_keys_len[tag_index] = tag_key_len;
+
+            /* allocate new memory for the tag key value */
+            /* + 1, because we need one more character for \0 */
+            ctx->tag_keys[tag_index] = flb_sds_create_size(tag_key_len + 1);
+            if (!ctx->tag_keys[tag_index]) {
+                flb_errno();
+                flb_sds_destroy(tags_list);
+                return -2;
+            }
+
+            /* tag_key points to the first character of tag key as char* */
+            tag_key = tags_list + tag_start;
+            strcpy(ctx->tag_keys[tag_index], tag_key);
+
+            flb_plg_debug(ctx->ins, "tag found: %s (len=%d)", ctx->tag_keys[tag_index],
+                          tag_key_len);
+
+            tag_index++;
+            tag_start = tag_end + 1;
+        }
+        tag_end++;
+    }
+
+    flb_sds_destroy(tags_list);
+
+    return ret;
+}
+
+/* Get EC2 instance tag values from /latest/meta-data/tags/instance/{tag_key}.
+ * Initializes ctx->tag_values and ctx->tag_values_len.
+ */
+static int get_ec2_tag_values(struct flb_filter_aws *ctx)
+{
+    int ret;
+    size_t i;
+    flb_sds_t tag_value = NULL;
+    size_t tag_value_len = 0;
+    size_t tag_value_path_len;
+    flb_sds_t tag_value_path;
+
+    /* initialize array for the tag values */
+    ctx->tag_values = flb_calloc(ctx->tags_count, sizeof(flb_sds_t*));
+    if (!ctx->tag_values) {
+        flb_errno();
+        return -1;
+    }
+    ctx->tag_values_len = flb_calloc(ctx->tags_count, sizeof(size_t*));
```

Comment:
I think we just need size_t here, not a pointer

Reply:
Also, for both of these calloc calls for a pointer on the context, do we need to check if it was already allocated and free the existing pointer? Alternative is to add error handling in this function to free them. Imagine you get a failure down below, then this will be already allocated and then it will be re-allocated when a retry occurs, right?

Reply:
> I think we just need size_t here, not a pointer

Fixed, https://github.com/fluent/fluent-bit/pull/6065/commits/88ffc82a259bbde1dd9c5a0cd8ad445655ac988a

> for both of these calloc calls for a pointer on the context, do we need to check if it was already allocated and free the existing pointer?

I guess, we don't need to.

At the beginning of `get_ec2_tags` we always free our memory related to tags and only then execute `get_ec2_tag_keys` and `get_ec2_tag_values`. Thus the variables should be always `NULL` and we should re-allocate memory for them.

```c
static int get_ec2_tags(struct flb_filter_aws *ctx)
{
    int ret;

    /* get_ec2_tags function might be called multiple times, so we need to always */
    /* free memory for tags in case of previous allocations */
    flb_filter_aws_tags_destroy(ctx);
    
    // ...
```

> Imagine you get a failure down below, then this will be already allocated and then it will be re-allocated when a retry occurs, right?

If the failure will happen, then we would call `get_ec2_tags` and free the memory (and set corresponding values to NULL in ctx). After destroy we would proceed with fetching tag keys and values, but inside these functions we would still have to allocate new chunks of memory.

Does it seem to be a correct implementation?

Reply:
Makes sense. Thanks!